File: js-agent/src/LedgerApi.ts
import { Agent, AnonymousIdentity, HttpAgent, SignIdentity } from "@dfinity/agent";
import ledgerBuilder from "./canisters/ledger/builder";
import LedgerService, {
    AccountIdentifier,
    BlockHeight,
    GetBalancesRequest,
    E8s,
    SendICPTsRequest
} from "./canisters/ledger/model";
import ledgerViewBuilder from "./canisters/nnsUI/builder";
import governanceBuilder from "./canisters/governance/builder";
import LedgerViewService, {
    CreateSubAccountResponse,
    GetTransactionsRequest,
    GetTransactionsResponse,
    NamedSubAccount
} from "./canisters/nnsUI/model";
import { create_dummy_proposals, test_happy_path } from "./tests";
import createNeuronImpl, { CreateNeuronRequest } from "./canisters/ledger/createNeuron";
import GovernanceService, { NeuronId } from "./canisters/governance/model";

export default class LedgerApi {
    private readonly ledgerService: LedgerService;
    private readonly ledgerViewService: LedgerViewService;
    private readonly governanceService: GovernanceService;
    private readonly agent: Agent;
    private readonly host: string;
    private readonly identity: SignIdentity;

    constructor(host: string, identity: SignIdentity) {
        const agent = new HttpAgent({
            host,
            identity
        });
        this.ledgerService = ledgerBuilder(agent, identity);
        this.ledgerViewService = ledgerViewBuilder(agent);
        this.governanceService = governanceBuilder(agent, identity, this.ledgerViewService.syncTransactions);
        this.agent = agent;
        this.host = host;
        this.identity = identity;
    }

    // Temporary method for demo purposes only, to give the specified account some ICPTs
    // by sending from the anon account which has been gifted lots of ICPTs
    public acquireICPTs = async (accountIdentifier: AccountIdentifier, e8s: E8s): Promise<void> => {
        const anonIdentity = new AnonymousIdentity();
        const agent = new HttpAgent({
            host: this.host,
            identity: anonIdentity
        });
        const anonLedgerService = ledgerBuilder(agent, anonIdentity);
        await anonLedgerService.sendICPTs({
            to: accountIdentifier,
            amount: e8s
        });
        await this.ledgerViewService.syncTransactions();
    }

    public createSubAccount = (name: string) : Promise<CreateSubAccountResponse> => {
        return this.ledgerViewService.createSubAccount(name);
    }

    public getAccount = async () : Promise<AccountDetails> => {
        const response = await this.ledgerViewService.getAccount();
        if ("Ok" in response) {
            return response.Ok;
        } else {
            const accountIdentifier = await this.ledgerViewService.addAccount();
            return {
                accountIdentifier,
                subAccounts: []
            };
        }
    }

    public getBalances = (request: GetBalancesRequest) : Promise<Record<AccountIdentifier, E8s>> => {
        return this.ledgerService.getBalances(request);
    }

    public getTransactions = (request: GetTransactionsRequest) : Promise<GetTransactionsResponse> => {
        return this.ledgerViewService.getTransactions(request);
    }

    public sendICPTs = async (request: SendICPTsRequest): Promise<BlockHeight> => {
        const response = await this.ledgerService.sendICPTs(request);
        this.ledgerViewService.syncTransactions();
        return response;
    }

    public createNeuron = async (request: CreateNeuronRequest) : Promise<NeuronId> => {
        return createNeuronImpl(
            this.identity, 
            this.ledgerService, 
            this.governanceService, 
            request);
    }

    public integrationTest = async (): Promise<void> => {
        return await test_happy_path(this.host, this.identity);
    }

    public createDummyProposals = async (neuronId: bigint): Promise<void> => {
        return await create_dummy_proposals(this.host, this.identity, neuronId);
    }
}

export type AccountDetails = {
    accountIdentifier: AccountIdentifier,
    subAccounts: Array<NamedSubAccount>
}

File: dfinity_wallet/lib/ic_api/web/ledger_api.dart
@JS()
library dfinity_agent.js;

import 'package:dfinity_wallet/ic_api/web/web_ic_api.dart';
import 'package:js/js.dart';

import '../models.dart';
import 'js_utils.dart';
import 'js_utils.dart';
import 'js_utils.dart';

@JS('createLedgerApi')
external LedgerApi createLedgerApi(String host, dynamic identity);

@JS("LedgerApi")
class LedgerApi {
  external factory LedgerApi(String host, dynamic identity);

  @JS("acquireICPTs")
  external Promise<void> acquireICPTs(String accountIdentifier, BigInt icpts);

  @JS("getAccount")
  external Promise<dynamic> getAccount();

  @JS("getBalances")
  external Promise<dynamic> getBalances(Object request);

  @JS("createSubAccount")
  external Promise<dynamic> createSubAccount(String name);

  @JS("sendICPTs")
  external Promise<dynamic> sendICPTs(Object request);

  @JS("createNeuron")
  external Promise<void> createNeuron(dynamic request);

  @JS("createCanister")
  external Promise<void> createCanister(dynamic request);

  @JS("topupCanister")
  external Promise<void> topupCanister(dynamic request);

  @JS("integrationTest")
  external Promise<void> integrationTest();

  @JS("getTransactions")
  external Promise<dynamic> getTransactions(dynamic request);

  @JS("createDummyProposals")
  external Promise<dynamic> createDummyProposals(String neuronId);
}


@JS("HardwareWalletApi")
class HardwareWalletApi {
  external factory HardwareWalletApi(String host, dynamic identity);

  @JS("sendICPTs")
  external Promise<void> sendICPTs (String fromAccount, SendICPTsRequest request);

  @JS("showAddressAndPubKeyOnDevice")
  external Promise<void> showAddressAndPubKeyOnDevice ();

}